plugins/datastreamer_plugin/fastdds/ReaderHandler.hpp

```diff
@@ -0,0 +1,145 @@
+
+#include <atomic>
+
+#include <fastdds/dds/topic/Topic.hpp>
+#include <fastdds/dds/subscriber/DataReader.hpp>
+#include <fastdds/dds/subscriber/DataReaderListener.hpp>
+
+#include "FastDdsListener.hpp"
+#include "utils/dynamic_types_utils.hpp"
+
+namespace eprosima {
+namespace plotjuggler {
+namespace fastdds {
+
+/**
+ * @brief TODO
+ *
+ */
+struct ReaderHandler : public eprosima::fastdds::dds::DataReaderListener
+{
+public:
+
+    ////////////////////////////////////////////////////
+    // CREATION & DESTRUCTION
+    ////////////////////////////////////////////////////
+
+    ReaderHandler(
+            eprosima::fastdds::dds::Topic* topic,
+            eprosima::fastdds::dds::DataReader* datareader,
+            eprosima::fastrtps::types::DynamicType_ptr type,
+            FastDdsListener* listener);
+
+
+    virtual ~ReaderHandler();
+
+    ReaderHandler& operator =(
+            ReaderHandler&& other);
+
+    ReaderHandler& operator =(
+            const ReaderHandler& other);
+
+    ////////////////////////////////////////////////////
+    // INTERACTION METHODS
+    ////////////////////////////////////////////////////
+
+    void stop();
+
+
+    ////////////////////////////////////////////////////
+    // LISTENER [ DATAREADER ] METHODS
+    ////////////////////////////////////////////////////
+
+    void on_data_available(
+            eprosima::fastdds::dds::DataReader* reader) override;
+
+
+    ////////////////////////////////////////////////////
+    // VALUES METHODS
+    ////////////////////////////////////////////////////
+
+    std::string topic_name() const;
+
+    std::string type_name() const;
+
+    std::vector<std::string> numeric_data_series_names() const;
+
+    std::vector<std::string> string_data_series_names() const;
+
+
+    ////////////////////////////////////////////////////
+    // AUXILIAR STATIC METHODS
+    ////////////////////////////////////////////////////
+
+    /**
+     * @brief Get default mask
+     *
+     * Callbacks accepted by this mask:
+     * - on_data_available
+     *
+     * @return eprosima::fastdds::dds::StatusMask with callbacks needed
+     */
+    static eprosima::fastdds::dds::StatusMask default_listener_mask_();
+
+
+    ////////////////////////////////////////////////////
+    // INTERNAL VARIABLES
+    ////////////////////////////////////////////////////
+
+    FastDdsListener* listener_;
+
+
+    ////////////////////////////////////////////////////
+    // FAST DDS POINTERS
+    ////////////////////////////////////////////////////
+
+    //! Topic related with this DataReader
+    eprosima::fastdds::dds::Topic* topic_;
+
+    //! DataReader
+    eprosima::fastdds::dds::DataReader* reader_;
+
+    //! Type Informantion
+    eprosima::fastrtps::types::DynamicType_ptr type_;
+
+    //! Data Type element
+    eprosima::fastrtps::types::DynamicData_ptr data_;
+
+    std::atomic<bool> stop_;
```

Comment:
This it is required indeed (at least I think so) because `stop` will be called from Qt thread, while `on_data_available` will be called from reading thread of Fast DDS, and both get/set this value.

Reply:
Ok, not important here because of deadline issues, but very important in the future: when dealing with atomics we should be more expressive and careful to avoid future headaches / endless debugging and to better express the intent behind the underlying synchronization: use member functions load, store, with explicit memory orderings, mentioning which data is meant to be synchronized, if any.